libindy/src/api/anoncreds.rs

```diff
@@ -926,3 +1010,118 @@ pub extern fn indy_verifier_verify_proof(command_handle: i32,
+                                  cb: Option<extern fn(
+                                      xcommand_handle: i32, err: ErrorCode,
+                                      witness_json: *const c_char
+                                  )>) -> ErrorCode {
+    check_useful_c_str!(rev_reg_def_json, ErrorCode::CommonInvalidParam4);
+    check_useful_c_str!(rev_reg_delta_json, ErrorCode::CommonInvalidParam5);
+    check_useful_c_callback!(cb, ErrorCode::CommonInvalidParam7);
+
+    let result = CommandExecutor::instance()
+        .send(Command::Anoncreds(AnoncredsCommand::Prover(ProverCommand::CreateWitness(
+            wallet_handle,
+            tails_reader_handle,
+            rev_reg_def_json,
+            rev_reg_delta_json,
+            rev_idx,
+            Box::new(move |result| {
+                let (err, witness_json) = result_to_err_code_1!(result, String::new());
+                let witness_json = CStringUtils::string_to_cstring(witness_json);
+                cb(command_handle, err, witness_json.as_ptr())
+            })
+        ))));
+
+    result_to_err_code!(result)
+}
+
+#[no_mangle]
+pub extern fn indy_update_witness(command_handle: i32,
+                                  wallet_handle: i32,
+                                  tails_reader_handle: i32,
+                                  witness_json: *const c_char,
+                                  rev_reg_def_json: *const c_char,
+                                  rev_reg_delta_json: *const c_char,
+                                  rev_idx: u32,
+                                  cb: Option<extern fn(
+                                      xcommand_handle: i32, err: ErrorCode,
+                                      updated_witness_json: *const c_char
+                                  )>) -> ErrorCode {
+    check_useful_c_str!(witness_json, ErrorCode::CommonInvalidParam4);
+    check_useful_c_str!(rev_reg_def_json, ErrorCode::CommonInvalidParam5);
+    check_useful_c_str!(rev_reg_delta_json, ErrorCode::CommonInvalidParam6);
+    check_useful_c_callback!(cb, ErrorCode::CommonInvalidParam8);
+
+    let result = CommandExecutor::instance()
+        .send(Command::Anoncreds(AnoncredsCommand::Prover(ProverCommand::UpdateWitness(
+            wallet_handle,
+            tails_reader_handle,
+            witness_json,
+            rev_reg_def_json,
+            rev_reg_delta_json,
+            rev_idx,
+            Box::new(move |result| {
+                let (err, updated_witness_json) = result_to_err_code_1!(result, String::new());
+                let updated_witness_json = CStringUtils::string_to_cstring(updated_witness_json);
+                cb(command_handle, err, updated_witness_json.as_ptr())
+            })
+        ))));
+
+    result_to_err_code!(result)
+}
+
+#[no_mangle]
+pub extern fn indy_store_witness(command_handle: i32,
+                                 wallet_handle: i32,
+                                 id: *const c_char,
```

Comment:
What is id here?

Reply:
Witness and Claim must be associated. We decided to pass the same id as a parameter in store_witness and store_credential

Reply:
But in theory we can have multiple witnesses for one claim for different timestamps.